// @nuxtjs/seo/dist/const.cjs
'use strict';

const NuxtSEO = {
  slug: "nuxt-seo",
  label: "Nuxt SEO",
  icon: "i-carbon-3rd-party-connected",
  description: "The all-in-one module that brings it all together.",
  repo: "harlan-zw/nuxt-seo",
  npm: "@nuxtjs/seo",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-zycxux?file=public%2F_robots.txt",
    i18n: "https://stackblitz.com/edit/nuxt-starter-pnej8lvb?file=public%2F_robots.txt"
  }
};
const SiteConfigModule = {
  slug: "site-config",
  label: "Site Config",
  icon: "i-carbon-settings-check",
  description: "Powerful build and runtime shared site configuration for Nuxt modules.",
  repo: "harlan-zw/nuxt-site-config",
  npm: "nuxt-site-config",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-zycxux?file=public%2F_robots.txt"
  }
};
const RobotsModule = {
  slug: "robots",
  label: "Robots",
  icon: "i-carbon-bot",
  description: "Tame the robots crawling and indexing your site with ease.",
  repo: "nuxt-modules/robots",
  npm: "@nuxtjs/robots",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-zycxux?file=public%2F_robots.txt",
    i18n: "https://stackblitz.com/edit/nuxt-starter-pnej8lvb?file=public%2F_robots.txt"
  }
};
const SitemapModule = {
  slug: "sitemap",
  label: "Sitemap",
  icon: "i-carbon-load-balancer-application",
  description: "Powerfully flexible XML Sitemaps that integrate seamlessly.",
  repo: "nuxt-modules/sitemap",
  npm: "@nuxtjs/sitemap",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-dyraxc?file=server%2Fapi%2F_sitemap-urls.ts",
    i18n: "https://stackblitz.com/edit/nuxt-starter-jwuie4?file=app.vue",
    manualChunking: "https://stackblitz.com/edit/nuxt-starter-umyso3?file=nuxt.config.ts",
    nuxtContent: "https://stackblitz.com/edit/nuxt-starter-a5qk3s?file=nuxt.config.ts"
  }
};
const OgImageModule = {
  slug: "og-image",
  label: "OG Image",
  icon: "i-carbon-image-search",
  description: "Generate OG Images with Vue templates in Nuxt.",
  repo: "nuxt-modules/og-image",
  npm: "nuxt-og-image",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-pxs3wk?file=pages/index.vue",
    i18n: "https://stackblitz.com/edit/nuxt-starter-uw7pqmxg?file=nuxt.config.ts",
    nuxtContent2: "https://stackblitz.com/edit/github-hgunsf?file=package.json",
    nuxtContent3: "https://stackblitz.com/edit/github-hgunsf-wd8esdec"
  }
};
const LinkCheckerModule = {
  slug: "link-checker",
  label: "Link Checker",
  icon: "i-carbon-cloud-satellite-link",
  description: "Find and magically fix links that may be negatively effecting your SEO.",
  repo: "harlan-zw/nuxt-link-checker",
  npm: "nuxt-link-checker",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-r2wzt1?file=nuxt.config.ts"
  }
};
const SeoUtilsModule = {
  slug: "seo-utils",
  label: "SEO Utils",
  icon: "i-carbon-tools",
  description: "SEO utilities to improve your Nuxt sites discoverability and shareability.",
  repo: "harlan-zw/nuxt-seo-utils",
  npm: "nuxt-seo-utils",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-vbay3q?file=app.vue"
  }
};
const SchemaOrgModule = {
  slug: "schema-org",
  label: "Schema.org",
  icon: "i-carbon-chart-relationship",
  description: "The quickest and easiest way to build Schema.org graphs.",
  repo: "harlan-zw/nuxt-schema-org",
  npm: "nuxt-schema-org",
  playgrounds: {
    basic: "https://stackblitz.com/edit/nuxt-starter-z9np1t?file=app.vue"
  }
};
const modules = [
  NuxtSEO,
  RobotsModule,
  SitemapModule,
  OgImageModule,
  SchemaOrgModule,
  LinkCheckerModule,
  SeoUtilsModule,
  SiteConfigModule
];
const explicitModules = [
  RobotsModule,
  SitemapModule,
  OgImageModule,
  SchemaOrgModule,
  LinkCheckerModule,
  SeoUtilsModule
];

exports.LinkCheckerModule = LinkCheckerModule;
exports.NuxtSEO = NuxtSEO;
exports.OgImageModule = OgImageModule;
exports.RobotsModule = RobotsModule;
exports.SchemaOrgModule = SchemaOrgModule;
exports.SeoUtilsModule = SeoUtilsModule;
exports.SiteConfigModule = SiteConfigModule;
exports.SitemapModule = SitemapModule;
exports.explicitModules = explicitModules;
exports.modules = modules;
