// @unhead/schema-org/dist/svelte.mjs
import { useUnhead, useHead } from '@unhead/svelte';
import { s as schemaAutoImports } from './shared/schema-org.q010LqYD.mjs';
import { a0 as UnheadSchemaOrg, aC as normalizeSchemaOrgInput } from './shared/schema-org.DoKdjnca.mjs';
export { a6 as defineArticle, at as defineBook, a7 as defineBreadcrumb, a8 as defineComment, au as defineCourse, a9 as defineEvent, aa as defineFoodEstablishment, ad as defineHowTo, af as defineImage, av as defineItemList, ag as defineJobPosting, ah as defineLocalBusiness, ax as defineMovie, ak as defineOrganization, al as definePerson, am as defineProduct, an as defineQuestion, ao as defineRecipe, ap as defineReview, aA as defineSoftwareApp, aq as defineVideo, ar as defineWebPage, as as defineWebSite } from './shared/schema-org.DoKdjnca.mjs';
import 'defu';
import 'unhead/plugins';
import 'unhead/utils';
import 'ufo';
import 'ohash';

function useSchemaOrg(input = [], options = {}) {
  const unhead = options.head || useUnhead();
  unhead.use(UnheadSchemaOrg());
  const entry = useHead(normalizeSchemaOrgInput(input), options);
  const corePatch = entry.patch;
  entry.patch = (input2) => corePatch(normalizeSchemaOrgInput(input2));
  return entry;
}
const schemaOrgAutoImports = [
  {
    from: "@unhead/schema-org/svelte",
    imports: schemaAutoImports
  }
];

export { schemaOrgAutoImports, useSchemaOrg };
